File: src/CTC/TrainSystem.cpp
/**
 * @file TrainSystem.cpp
 *
 * @brief Implementations of the TrainSystem class
 *
 * @author Andrew Toader
 *
 * @date 10.01.2020
 *
*/

#include "TrainSystem.hpp"  // Definition of class
#include <vector>           // For accessing list of Trains, Tracks, Signals
#include "Logger.hpp"       // For logging events (debugging purposes)
#include "CTCMain.hpp"
#include "SWTrackControllerMain.hpp"    // For sending information to the Train Controller

/**
 * @brief	gets singleton instance
 *
 * @return 	reference to this singleton TrainSystem Object
 *
 */
TrainSystem& TrainSystem::GetInstance()
{
	static TrainSystem* pInstance = new TrainSystem();
	return *(pInstance);
}

/**
* @brief Import the layout of the track
 *
* @param none
*
* @return None
*
*/
void TrainSystem::ImportTrackLayout()
{
	// Initialize the red track blocks
	Track* ptemp_track = nullptr;
	for(int i = 0; i < 76; i++)
	{
		ptemp_track = new Track();
		p_blocks_red.push_back(ptemp_track);
	}

	// Initialize the green track blocks
	for(int i = 0; i < 150; i++)
	{
		ptemp_track = new Track();
		p_blocks_green.push_back(ptemp_track);
	}
	ptemp_track = nullptr;

	// Create Green Switches
	Switch* sw1 = new Switch(1, 12);
	Switch* sw2 = new Switch(30, 150);
	Switch* sw3 = new Switch(-1, 59);
	Switch* sw4 = new Switch(-1, 61);
	Switch* sw5 = new Switch(76, 101);
	Switch* sw6 = new Switch(86, 100);
	p_switches_green.push_back(sw1);
	p_switches_green.push_back(sw2);
	p_switches_green.push_back(sw3);
	p_switches_green.push_back(sw4);
	p_switches_green.push_back(sw5);
	p_switches_green.push_back(sw6);

	// Create Red Switches
	Switch* rsw1 = new Switch(-1, 8);
	Switch* rsw2 = new Switch(1, 15);
	Switch* rsw3 = new Switch(28, 76);
	Switch* rsw4 = new Switch(32, 72);
	Switch* rsw5 = new Switch(39, 71);
	Switch* rsw6 = new Switch(43, 67);
	Switch* rsw7 = new Switch(53, 66);
	p_switches_red.push_back(rsw1);
	p_switches_red.push_back(rsw2);
	p_switches_red.push_back(rsw3);
	p_switches_red.push_back(rsw4);
	p_switches_red.push_back(rsw5);
	p_switches_red.push_back(rsw6);
	p_switches_red.push_back(rsw7);

	LOG_CTC("From TrainSystem::ImportTrackLayout() : Tracks Created");
}

/**
* @brief Get the Array of track pointers for the green track
 *
* @param[in] enum Line
*
* @return vector<Track*>
* @return Throw Error if incorrect line is specified
*
*/
std::vector<Track*> TrainSystem::GetTrackArr(enum Line ln)
{
	if(ln == LINE_GREEN)
	{
		return p_blocks_green;
	}
	else if(ln == LINE_RED)
	{
		return p_blocks_red;
	}
	else
	{
		throw std::logic_error("TrainSystem::GetTrackArr : Invalid Line Argument");
	}
}

/**
* @brief Get the Array of Signal pointers
*
* @param Line
*
* @return vector<Signal*>
*
*/
std::vector<Switch*> TrainSystem::GetSwitchesArr(enum Line ln)
{
	if(ln == LINE_GREEN)
	{
		return p_switches_green;
	}
	else if(ln == LINE_RED)
	{
		return p_switches_red;
	}
	else
	{
		throw std::logic_error("TrainSystem::GetTrackArr : Invalid Line Argument");
	}
}

/**
* @brief Get the Array of train pointers
*
* @param none
*
* @return vector<Train*>
*
*/
std::vector<Train*> TrainSystem::GetTrainArr()
{
	return p_trains;
}

/**
 *@brief Create(dispatch) a new train by creating
 * the Train object then adding it to the class member vector
 *
 * @param[in]	block_to
 *
 * @return pointer to newly created Train struct
 *
 */
Train* TrainSystem::CreateNewTrain(int block_to, enum Line ln)
{
	// Set the train number to the next available
	int num = p_trains.size() + 1;

	// Create an object with that number
	Train* p_temp;
	p_temp = new Train(num, block_to);

	// Append the train to the train list
	p_trains.push_back(p_temp);
	train_numbers.push_back(num);

	// Add Speed and Authority to train
	p_temp->authority = 1000;           // m
	p_temp->command_speed = 40;         // km/hr

	// Log creation of object
	LOG_CTC("From TrainSystem::CreateNewTrain : Created Train #%d", p_temp->train_id);

	// return the object just created
	return p_temp;
}

/**
* @brief change Track status to occupied
*
* @param[in]	track_num
* @param[in]	Line
*
* @return Throw Error if track is out of range or invalid line entered
*
*/
void TrainSystem::SetTrackOccupied(int track_num, enum Line ln)
{
	if(ln == LINE_GREEN)
	{
		// Make sure track is not out of range
		if ((track_num < 1) || track_num > p_blocks_green.size())
		{
			throw std::logic_error("TrainSystem::SetTrackOccupied() index out of bounds");
		}

		// Set occupied member variable as true
		p_blocks_green[track_num - 1]->occupied = true;
	}
	else if(ln == LINE_RED)
	{
		// Make sure track is not out of range
		if ((track_num < 1) || track_num > p_blocks_red.size())
		{
			throw std::logic_error("TrainSystem::SetTrackOccupied() : index out of bounds");
		}

		// Set occupied member variable as true
		p_blocks_red[track_num - 1]->occupied = true;
	}
	else
	{
		throw std::logic_error("TrainSystem::SetTrackOccupied() : Invalid input for Line");
	}
}

/**
* @brief change Track status to not occupied
*
* @param[in]	track_num
* @param[in]	Line
*
* @return Throw Error if track is out of range or invalid line entered
*
*/
void TrainSystem::SetTrackNotOccupied(int track_num, enum Line ln)
{
	if(ln == LINE_GREEN)
	{
		// Make sure track is not out of range
		if ((track_num < 1) || track_num > p_blocks_green.size())
		{
			throw std::logic_error("TrainSystem::SetTrackNotOccupied() index out of bounds");
		}

		// Set occupied member variable as true
		p_blocks_green[track_num - 1]->occupied = false;
	}
	else if(ln == LINE_RED)
	{
		// Make sure track is not out of range
		if ((track_num < 1) || track_num > p_blocks_red.size())
		{
			throw std::logic_error("TrainSystem::SetTrackNotOccupied() : index out of bounds");
		}

		// Set occupied member variable as true
		p_blocks_red[track_num - 1]->occupied = false;
	}
	else
	{
		throw std::logic_error("TrainSystem::SetTrackNotOccupied() : Invalid input for Line");
	}
}

/**
 * @brief change Track status to occupied
 *
 * @param[in]	switch_num, Line, update_pos
 *
 * @return none
 *
*/
void TrainSystem::SetSwitch(int switch_num, enum Line ln, int pos)
{
	if(ln == LINE_GREEN)
	{
		if(pos == 1)
		{
			p_switches_green[switch_num - 1]->pointing_to = p_switches_green[switch_num - 1]->greater_block;
		}
		else if(pos == 0)
		{
			p_switches_green[switch_num - 1]->pointing_to = p_switches_green[switch_num - 1]->less_block;
		}
		else
		{
			throw std::logic_error("TrainSystem::SetSwitch : Invalid position given (not 1 or 0)");
		}
	}
	if(ln == LINE_RED)
	{
		if(pos == 1)
		{
			p_switches_red[switch_num - 1]->pointing_to = p_switches_red[switch_num - 1]->greater_block;
		}
		else if(pos == 0)
		{
			p_switches_red[switch_num - 1]->pointing_to = p_switches_red[switch_num - 1]->less_block;
		}
		else
		{
			throw std::logic_error("TrainSystem::SetSwitch : Invalid position given (not 1 or 0)");
		}
	}
	else
	{
		throw std::logic_error("TrainSystem::SetSwitch : Invalid Line provided");
	}
}

/**
 * @brief update the train position (to be called every time occupancy is changed)
 *
 * @param[post] train location for each Train* in p_trains updated with current location
 *
 * @return none
 *
*/
void TrainSystem::UpdateTrainPosition()
{
	for(int i = 0; i < p_trains.size(); i++)
	{
		if(p_trains[i]->line_on == LINE_GREEN)      // If train is on GREEN Line
		{
			if(p_trains[i]->index_on_route == 0)        // Has not made it out of the yard yet
			{
				if(p_blocks_green[green_route_blocks[1] - 1]->occupied == false)
				{
					p_trains[i]->index_on_route++;
					p_trains[i]->authority--;
				}
				else
				{
					continue;
				}
			}
			else if(p_trains[i]->index_on_route == (green_route_blocks.size() - 1))     // Reached the yard (end)
			{
				// Free the train and delete it from the vector
				delete p_trains[i];
				p_trains.erase(p_trains.begin() + i);
				i--;
				continue;
			}
			else
			{
				if(p_blocks_green[p_trains[i]->index_on_route - 1]->occupied == false)
				{
					p_trains[i]->index_on_route++;
					p_trains[i]->authority--;
				}
				else
				{
					continue;
				}
			}
		}
		else        // If on RED Line
		{
			if (p_trains[i]->index_on_route == 0)        // Has not made it out of the yard yet
			{
				if (p_blocks_red[red_route_blocks[1] - 1]->occupied == false)
				{
					p_trains[i]->index_on_route++;
					p_trains[i]->authority--;
				}
				else
				{
					continue;
				}
			}
			else if (p_trains[i]->index_on_route == (red_route_blocks.size() - 1))     // Reached the yard (end)
			{
				// Free the train and delete it from the vector
				delete p_trains[i];
				p_trains.erase(p_trains.begin() + i);
				i--;
				continue;
			}
			else
			{
				if (p_blocks_red[p_trains[i]->index_on_route - 1]->occupied == false)
				{
					p_trains[i]->index_on_route++;
					p_trains[i]->authority--;
				}
				else
				{
					continue;
				}
			}
		}

		// Update Authority if needed
		if(p_trains[i]->authority <= 1)
		{
			// Update internal variables
			p_trains[i]->authority = 3;

			// Send to Track Controller
			Common::Request reqSend;
			reqSend.SetRequestCode(Common::RequestCode::SWTRACK_UPDATE_AUTHORITY);
			reqSend.SetData(std::to_string(i));
			reqSend.SetData(std::to_string(3));
			SWTrackController::serviceQueue.Push(reqSend);  // Push request to SW Track Controller Queue

			LOG_CTC("Authority for train %d updated and sent to SWTC", i + 1);
		}
	}
}

File: src/CTC/TrainSystem.py
"""
	@file TrainSystem.hpp

	@brief Declaration of the TrainSystem class; will run the CTC module

	@author Andrew Toader

	@date 10.01.2020
"""
import sys

from src.CTC.CTCDef import *

"""
	@class TrainSystem

	@brief Singleton class responsible for running vital operations of the CTC
	
	@members 
		@li train_numbers
		@li trains_arr
		@li blocks_green_arr
		@li blocks_red_arr
		@li switches_green_arr
		@li switches_red_arr
		@li green_route_blocks
		@li green_route_switches
		@li red_route_blocks
		@li red_route_switches
		
	@methods
		@li ImportTrackLayout
		@li DispatchTrain
		@li ReturnOccupancies
		@li ReturnSwitchPositions
		
"""


class TrainSystem:
	def __init__(self):
		self.ImportTrackLayout()
		self.next_train_num = 1
		self.train_numbers = []
		self.trains_arr = []
		self.green_route_blocks = [-1, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81,
		                           82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 85, 84,
		                           83, 82, 81, 80, 79, 78, 77, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111,
		                           112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128,
		                           129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145,
		                           146, 147, 148, 149, 150, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15,
		                           14, 13, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
		                           22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42,
		                           43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, -1]
		self.green_route_switches = [0, 0, 0, 1, 1, 1, 0, 1, 0, 0]
		self.red_route_blocks = [-1, 9, 8, 7, 6, 5, 4, 3, 2, 1, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 76, 75,
		                         74, 73, 72, 33, 34, 35, 36, 37, 38, 71, 70, 69, 68, 67, 44, 45, 46, 47, 48, 49, 50, 51,
		                         52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 52, 51, 50, 49, 48, 47, 46,
		                         45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24,
		                         23, 22, 21, 20, 19, 18, 17, 16, 1, 2, 3, 4, 5, 6, 7, 8, -1]
		self.red_route_switches = [0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1]

	"""
		@brief Import the layout of the track
		
		@param[out] updates green and red blocks and switches
		
		@return None
	"""

	def ImportTrackLayout(self):
		self.blocks_green_arr = []
		self.blocks_red_arr = []
		self.switches_green_arr = []
		self.switches_red_arr = []

		# Initialize red blocks
		for i in range(76):
			self.blocks_red_arr.append(Track())
		# Initialize green blocks
		for i in range(150):
			self.blocks_green_arr.append(Track())

		# Create Green Switches
		sw1 = Switch(1, 12)
		sw2 = Switch(30, 150)
		sw3 = Switch(-1, 59)
		sw4 = Switch(-1, 61)
		sw5 = Switch(76, 101)
		sw6 = Switch(86, 100)
		self.switches_green_arr = [sw1, sw2, sw3, sw4, sw5, sw6]

		# Create Red Switches
		rsw1 = Switch(-1, 8)
		rsw2 = Switch(1, 15)
		rsw3 = Switch(28, 76)
		rsw4 = Switch(32, 72)
		rsw5 = Switch(39, 71)
		rsw6 = Switch(43, 67)
		rsw7 = Switch(53, 66)
		self.switches_red_arr = [rsw1, rsw2, rsw3, rsw4, rsw5, rsw6, rsw7]

	"""
		@brief Create(dispatch) a new train by creating the Train object then adding it to the class member vector

		@param[in]	block_to
		@param[in]  line_on

		@return pointer to newly created Train struct
	"""

	def DispatchTrain(self, block_to, line):
		# Create the train which will dispatch
		temp_train = Train(self.next_train_num, block_to, line)

		# Give Train Speed and Authority (blocks, km/hr)
		temp_train.authority = 3
		temp_train.command_speed = 55

		# Add train to this array
		self.trains_arr.append(temp_train)
		self.train_numbers.append(self.next_train_num)

		# Increment the next train number counter
		self.next_train_num += 1

	"""
		@breif Function which returns an array of track occupancies which the CTC GUI can use to dispaly on the screen
		
		@param[in]  line to return track occupancies
		
		@return array of boolean values
	"""

	def ReturnOccupancies(self, line):
		to_send = []
		if line == Line.LINE_GREEN:
			for i in range(len(self.blocks_green_arr)):
				to_send.append(self.blocks_green_arr[i].occupied)
		elif line == Line.LINE_RED:
			for i in range(len(self.blocks_red_arr)):
				to_send.append(self.blocks_red_arr[i].occupied)
		else:
			raise Exception('CTC : TrainSystem.ReturnOccupancies recieved an erronious input')
		return to_send

	"""
		@breif Function which returns an array of track occupancies which the CTC GUI can use to dispaly on the screen

		@param[in]  line to return track occupancies

		@return array of boolean values
	"""

	def ReturnSwitchPositions(self, line):
		to_send = []
		if line == Line.LINE_GREEN:
			for i in range(len(self.switches_green_arr)):
				to_send.append(self.switches_green_arr[i].TrackSwitchToString())
		elif line == Line.LINE_RED:
			for i in range(len(self.switches_red_arr)):
				to_send.append(self.switches_red_arr[i].TrackSwitchToString())
		else:
			raise Exception('CTC : TrainSystem.ReturnSwitchPositions recieved an erronious input')
		return to_send


# Define a TrainSystem object to use; acts as equivalent of singleton class
ctc = TrainSystem()
